### PR TITLE
feat(remote): support wildcards in remote values/secrets file selectors

### PR DESCRIPTION
**Pull request by hoppla20**
## Summary
This PR adds support for glob patterns (wildcards) in the file selector portion of remote URLs, allowing users to reference multiple files from a remote repository with a single pattern instead of listing each file individually.

## Key Changes

- **New `HasGlobPattern()` function** in `pkg/remote/remote.go`: Detects whether a remote URL contains glob metacharacters (`*`, `?`, `[`) in its file selector portion, distinguishing them from legitimate uses in query strings and IPv6 addresses.

- **Enhanced `Source` struct**: Added `HasSelector` field to track whether a URL contains an explicit `@` separator, which is required to distinguish between repository roots and wildcard patterns when parsing URLs.

- **Wildcard validation in `Remote.Fetch()`**: Added validation to reject unsupported wildcard patterns:
  - Plain `https://` and `s3://` URLs (which fetch single files)
  - Non-archive `s3::` references
  - URLs without explicit `@` selectors (where Dir/File inference is ambiguous)

- **Glob expansion in `Storage.resolveFile()`**: When a remote URL contains a glob pattern, the fetched directory is expanded using `filepath.Match` (same syntax as local values files), with results sorted and filtered to regular files only.

- **Updated test coverage**: 
  - New `TestHasGlobPattern()` covering edge cases (IPv6 hosts, placeholders, character classes)
  - New `TestRemote_Fetch_Wildcard()` verifying cache behavior with wildcards
  - New `TestRemote_Fetch_WildcardUnsupported()` validating error cases
  - New `TestStorage_resolveFile_RemoteGlob()` testing glob expansion hermetically
  - Updated `TestParse()` to verify `HasSelector` field

- **Documentation updates** in `docs/environments.md`: Added examples and limitations for wildcard support in both environment secrets and values files.

## Implementation Details

- Wildcard expansion happens at the Storage layer (after fetch), not in Remote.Fetch, allowing the entire repository to be cached once and reused for different glob patterns.
- The cache key is independent of the file selector, so `dir/*.yaml` and `dir/specific.yaml` share the same cache entry.
- Only single-level glob matching is supported (no recursive `**`), consistent with local values file globbing.
- Results are sorted for deterministic ordering, matching local glob behavior.

### CODE REVIEW

**Comment by yxxhero** [comment]
@hoppla20 DCO issue.

**Comment by hoppla20** [comment]
Oh sorry, I will amend it

**Review by yxxhero — COMMENTED** [review]
Thanks for this PR! I did a deep review (read all changed files plus their consumers, ran `go build`, `go vet`, `go test ./pkg/remote/ ./pkg/state/` — all green, and `golangci-lint` shows no new findings).

## Overall

High-quality change. The layering is clean: detection (`remote.HasGlobPattern`) → validation (`Remote.Fetch` rejects shapes that can never expand) → expansion (`Storage.resolveFile` globs the fetched cache dir). The tricky bits — `?` being the query-string delimiter, `[` legitimately appearing in IPv6 hosts and `[$GITHUB_ORG]`-style placeholders, literal-file-before-glob — are all documented with clear rationale in comments.

## What I particularly like

- **Cache design**: the cache key derives from scheme/host/dir/query only, so `dir/*.yaml` and `dir/a.yaml` share one clone of the repo. `TestRemote_Fetch_Wildcard`'s cache-hit case locks this in.
- **Consistency with local globs**: same engine (`fs.Glob` → `filepath.Glob`/`filepath.Match`), same sort order, and the `FileExistsAt` filter keeps the same "regular files only" contract (`Mode().IsRegular()` excludes directory matches), matching `ExpandPaths`.
- **Precise rejection matrix in `Fetch`**: plain http(s)/s3 (single object), non-archive forced `s3::`, and no explicit `@` selector each get an actionable error message; the `s3::` archive + explicit-selector exception (`stripSubdirSelector` + decompressor branch) checks out.
- **Backwards compatible**: before this PR a wildcard selector always ended at `values file matching ... does not exist` (or a handler skip); now a single match works and a multi-match fails with a clearer message.
- **Good test hygiene**: `TestStorage_resolveFile_RemoteGlob` covers sort order, non-recursiveness, extension filtering, and the missing-file-handler matrix hermetically; the network test is pinned to a tag, consistent with the existing suite.
- `case fetchedFilePath == ""` correctly preserves the `ignoreMissingGitBranch` behavior (empty result falls through to the missing-file handler), matching the pre-PR `FileExistsAt("") == false` behavior.

## Findings

Inline below. Summary of the non-nit ones:

1. **Docs**: a raw `?` in the selector can never work (it starts the URL query string, silently dropping `?ref=...`); only `%3F` does. Worth a caveat in the limitations list.
2. **Scope**: the feature is fully functional only for environment `values:`/`secrets:`. Four release-path consumers still hard-error on `len(paths) > 1` (`state.go` `generateTemporaryReleaseValuesFilesCore`, `generateSecretValuesFiles`, `resolveReleaseValues`, `LoadYAMLForEmbedding`), while `create.go` `loadEnvValues` and `envvals_loader.go` iterate all matches. Since the PR title says "values/secrets" broadly and the docs only cover environments, a one-line note about the release-scope restriction (or a follow-up issue) would set the right expectations.
3. **Behavior corner**: an unresolvable selector containing an unclosed `[` now surfaces `filepath.ErrBadPattern` as a hard error even under `missingFileHandler: Info/Warn`, where it used to be treated as missing. Probably fine, but worth a conscious decision.

Minor: the commit message says "Fixed an existing err-shadowing hazard in the same code path" — `fetchedFilePath, err := r.Fetch(...)` still shadows the outer `err`; the restructure actually introduced a distinct `globErr` for the new Glob call. No functional impact (the outer `err` only matters for the local-path branch), just noting for the record.

Happy to see this merged once the docs nits are addressed — none of the findings are blockers from my side.

File: docs/environments.md [hunk 1]
```diff
@@ -261,6 +269,24 @@ For more information about the supported protocols see: [go-getter Protocol-Spec
...
+environments:
+  production:
+    values:
+      - git::https://github.com/org/repo.git@config/production/*.yaml?ref=main
+```
+
+All matches are fetched from a single clone of the repository, then merged in the same sorted order as a local `values:` glob (see [Precedence](#environment-values-precedence)).
+
+Notes and limitations:
+- The pattern is matched with the same syntax as local values file globs (`filepath.Match`: `*`, `?`, `[abc]`), against a single path segment — there is no recursive `**`.
```
Comment (yxxhero) [inline docs/environments.md:286]: Small but important docs caveat: a **raw `?` in the file selector can never work**. `url.Parse` splits the query string at the first `?`, so e.g. `@dir/v?.yaml?ref=main` parses as `File="v"` and `RawQuery=".yaml?ref=main"` — the ref is silently lost and the clone falls back to the default branch. `ParseQuery` doesn't even reject `.yaml?ref=main` (it becomes key `.yaml?ref`), so the failure is confusing rather than loud.

A `?` wildcard only works percent-encoded (`%3F`), which survives into the decoded `u.Path` and reaches the glob. Suggest adding a bullet like:

> A `?` wildcard must be percent-encoded as `%3F`, since a raw `?` starts the URL query string (e.g. `@dir/v%3F.yaml?ref=main`).

The code comment in `Remote.Fetch` already explains that `?` can never reach `u.File` — surfacing that to users here would prevent real confusion.

File: docs/environments.md [hunk 2]
```diff
@@ -261,6 +269,24 @@ For more information about the supported protocols see: [go-getter Protocol-Spec
...
+- Wildcards require an explicit `@` selector to mark the repository root, as in the example above.
+- Wildcards are only supported with getters that download a whole directory, such as `git::`. Plain `https://`/`s3://` references and a non-archive `s3::` reference each fetch a single file and cannot be expanded; use `git::` (or an `s3::` archive URL) instead.
```
Comment (yxxhero) [inline docs/environments.md:288]: Might be worth stating the **scope** of this feature explicitly here: since `resolveFile` now expands globs, a wildcard matching exactly one file also works in release-level `values:`/`secrets:`, but a **multi-match** wildcard still hard-errors there ("glob patterns in release values and secrets is not supported yet") in `generateTemporaryReleaseValuesFilesCore`, `generateSecretValuesFiles`, `resolveReleaseValues`, and `LoadYAMLForEmbedding`. Only environment values/secrets iterate over all matches.

The commit message acknowledges this ("Release values/secrets keep their existing ... restriction"), but the PR title ("remote values/secrets file selectors") and this docs section don't. A one-line note — and/or a follow-up issue, since the existing error message even says "please submit a feature request if necessary" — would set the right expectations.

File: pkg/state/storage.go [hunk 3]
```diff
@@ -69,8 +69,38 @@ func (st *Storage) resolveFile(missingFileHandler *string, tpe, path string, opt
...
+			// here, against the fetched directory, using the same glob syntax as
+			// local values files (st.ExpandPaths / filepath.Match).
+			//
+			// st.ExpandPaths itself is not reused: its normalizePath would
+			// incorrectly prefix the helmfile's basePath onto this
+			// already-absolute cache path whenever remote.CacheDir() falls back
+			// to the relative ".helmfile" directory.
+			matches, globErr := st.fs.Glob(fetchedFilePath)
+			if globErr != nil {
+				return nil, false, fmt.Errorf("failed processing %s: %v", path, globErr)
```
Comment (yxxhero) [inline pkg/state/storage.go:94]: Corner-case behavior change worth a conscious decision: if the selector contains an unclosed `[` (a legal filename character — e.g. a literal remote file `values-[foo.yaml` that doesn't exist at the fetched ref), `filepath.Glob` returns `ErrBadPattern` and this returns a hard `syntax error in pattern` failure — even when `missingFileHandler` is `Info`/`Warn`/`Debug`, whereas pre-PR the same input was treated as a missing file and skipped.

If that's acceptable, fine (it's genuinely rare); if not, something like:

```go
if globErr != nil && !errors.Is(globErr, fs.ErrBadPattern) {
	return nil, false, fmt.Errorf("failed processing %s: %v", path, globErr)
}
```

would degrade malformed patterns back to "no match" and let the handler decide. Either way, the error text could include `globErr` context that the pattern came from a remote selector (it already includes the original `path`, which helps).

File: pkg/remote/remote.go [hunk 4]
```diff
@@ -112,6 +119,30 @@ func IsRemote(goGetterSrc string) bool {
 	return true
 }
 
+// hasGlobMeta reports whether p contains a filepath.Match metacharacter. The
+// set matches filepath.Match's, which also backs local values-file globbing
+// (Storage.ExpandPaths), so remote and local glob syntax stay identical.
+func hasGlobMeta(p string) bool {
+	return strings.ContainsAny(p, `*?[`)
```
Comment (yxxhero) [inline pkg/remote/remote.go:126]: Nit: the `?` in `ContainsAny` can never trigger here for `u.File` — `url.Parse` splits the query at the first `?`, so `HasGlobPattern`'s `hasGlobMeta(u.File)` effectively only ever sees `*`/`[`. That makes this subtly different from `Fetch`'s `strings.Contains(u.File, "*")`, which is intentional (per the comment in `Fetch`, only `*` is treated as an unambiguous rejection signal, since `[` can appear in literal filenames).

Keeping `?` for symmetry with `filepath.Match`'s metachar set is fine as belt-and-braces (e.g. if a `%3F`-decoded selector ever reaches this path — which it does, and then it *does* matter!). Maybe just add a word here noting `?` only arrives via percent-encoded selectors, so future readers don't puzzle over the asymmetry.

File: pkg/remote/remote_test.go [hunk 5]
```diff
@@ -987,6 +1053,141 @@ func TestRemote_Fetch(t *testing.T) {
...
+				t.Errorf("unexpected result: unexpected cache miss")
+			}
+			if !tt.expectCacheHit && hit {
+				t.Errorf("unexpected result: unexpected cache hit")
+			}
+		})
+	}
+}
+
+func TestRemote_Fetch_WildcardUnsupported(t *testing.T) {
```
Comment (yxxhero) [inline pkg/remote/remote_test.go:1130]: Nice rejection matrix. One gap: there's no test for the *allowed* wildcard + `s3::` archive shape (the `decompressorForFile(filepath.Base(u.Dir)) != nil` exception). A hermetic sibling case — e.g. `s3::https://bucket.s3.us-east-2.amazonaws.com/dir/app.tar.gz@*.yaml` asserting that `Fetch` proceeds and the S3 getter is invoked — would lock in that exception so a future refactor of the validation doesn't silently break it (the same way the four cases above lock in the rejections).

**Review by yxxhero — APPROVED** [review]
